client/blocks/app-banner/index.jsx

```diff
@@ -131,14 +131,38 @@ class AppBanner extends Component {
 			}
 		}
 
-		//TODO: update when deferred deep links are available
 		if ( this.isiOS() ) {
-			return 'itms://itunes.apple.com/us/app/wordpress/id335703880?mt=8';
+			return this.getiOSDeepLink( currentRoute, currentSection );
 		}
 
 		return null;
 	}
 
+	getiOSDeepLink( currentRoute, currentSection ) {
+		const hasRoute = currentRoute !== null && currentRoute !== '/';
+		let fragment;
+
+		switch ( currentSection ) {
+			case EDITOR:
+				fragment = '/post';
+				break;
+			case NOTES:
+				fragment = '/notifications';
+				break;
+			case READER:
+				// The Reader is generally accessed at the root of WordPress.com ('/').
+				// In this case, we need to manually add the section name to the
+				// URL so that the iOS app knows which section to open.
+				fragment = hasRoute ? currentRoute : '/read';
+				break;
+			case STATS:
+				fragment = hasRoute ? currentRoute : '/stats';
+				break;
+		}
+
+		return `https://apps.wordpress.com/get#${ encodeURIComponent( fragment ) }`;
```

Comment:
note that we've exposed a case where we are using an unset variable - if `currentSection` isn't one of these four things then we'll be trying to encode `undefined`

Reply:
Thanks for catching that – I'm still in Swift mode where switches must be exhaustive and you'll get an error if they're not.

Reply:
With current code the `default` case will never be reached anyway, since `AppBanner` will bail early in render if one of those sections is not detected, see: https://github.com/Automattic/wp-calypso/blob/82f7c5fbd6905bc99a9638ea9075d7db48a14092/client/blocks/app-banner/index.jsx#L149 Still, that's not immediately obvious and it doesn't hurt to add safeguards here too. :+1: